torch_xla/csrc/module.cpp

```diff
@@ -574,6 +594,50 @@ void XlaModule::FlushTensorsOperations() {
   XLATensor::ApplyPendingGraph(tensors);
 }
 
+void XlaModule::SetBackwardSizeOpValues(
+    const std::unordered_map<size_t, XlaComputationInOut::ShapeSizes>&
+        ret_size_op_values) {
+  size_t backward_input_idx = 0;
+  for (const auto out_idx : df_input_vjps_) {
+    const auto ret_size_op_value_it = ret_size_op_values.find(out_idx);
+    if (ret_size_op_value_it != ret_size_op_values.end()) {
+      const auto it_ok = backward_size_op_values_.insert(
+          std::make_pair(backward_input_idx, ret_size_op_value_it->second));
+      XLA_CHECK(it_ok.second);
+    }
+    ++backward_input_idx;
+  }
+  backward_input_idx += df_input_captured_inputs_.size();
+  for (const auto out_idx : df_input_captured_outputs_) {
+    const auto ret_size_op_value_it = ret_size_op_values.find(out_idx);
+    if (ret_size_op_value_it != ret_size_op_values.end()) {
+      const auto it_ok = backward_size_op_values_.insert(
+          std::make_pair(backward_input_idx, ret_size_op_value_it->second));
+      XLA_CHECK(it_ok.second);
+    }
+    ++backward_input_idx;
+  }
+}
+
+void XlaModule::CheckAssumedSizes(
+    const TensorBatchVector::value_type& replica_raw_grad_outputs) {
+  for (const auto& kv : backward_size_op_values_) {
+    const auto& assumed_size = kv.second;
+    XLA_CHECK_LT(kv.first, replica_raw_grad_outputs.size());
+    const auto& raw_grad_output = replica_raw_grad_outputs[kv.first];
+    const auto effective_size = raw_grad_output->toTensor();
```

Comment:
Unfortunately, yes. Collecting all these is additional work I'd rather postpone for a future patch, especially since we rely on fusion for bigger networks anyway.

Reply:
I sit down and think we need to think this more carefully.
We need something smarter/better than having hundreds of single RPCs.

Reply:
We can add a multi version of `toTensor` in a followup.

Reply:
Yes, remove it altogether!